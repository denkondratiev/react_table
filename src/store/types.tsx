import {
  setParams,
  setTable,
  setRows,
  setCells,
  setShowButtons,
  increment,
  setNewRow,
  removeRow
} from './actions'

export const ACTION_TYPES = {
  SET_PARAMS: 'SET::PARAMS',
  SET_TABLE: 'SET::TABLE',
  SET_ROWS: 'SET::ROWS',
  SET_CELLS: 'SET::CELLS',
  SHOW_BUTTONS: 'SHOW::BUTTONS',
  INCREMENT: 'INCREMENT',
  ADD_ROW: 'ADD::ROW',
  REMOVE_ROW: 'REMOVE::ROW'
}

export type State = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
  params: Params
  buttons: boolean
}

export type Action = (
  | ReturnType<typeof setParams>
  | ReturnType<typeof setTable>
  | ReturnType<typeof setRows>
  | ReturnType<typeof setCells>
  | ReturnType<typeof setShowButtons>
  | ReturnType<typeof increment>
  | ReturnType<typeof setNewRow>
  | ReturnType<typeof removeRow>
)

export type Params = {
  [name: string]: string | number
}

export type RowsParams = {
  [name: string]: Array<string>
}

export type CellsParams = {
  [name: string]: { id: string, amount: number }
}

export type NewRowsParams = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
}
