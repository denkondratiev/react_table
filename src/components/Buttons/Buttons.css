.buttons {
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
}

.btn {
  max-width: 120px;
  margin-right: 5px;
}

.btn:last-child {
  margin-right: 0;
}
