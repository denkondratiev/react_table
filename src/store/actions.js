export const ACTION_TYPES = {
  SET_COLUMNS: 'SET::COLUMNS',
  SET_ROWS: 'SET::ROWS',
  SET_TABLE: 'SET::TABLE',
  SET_HIGHLIGHT: 'SET::HIGHLIGHT',
  SET_SHOW_TABLE: 'SET::SHOW::TABLE',
  SET_SHOW_BUTTONS: 'SET::SHOW::BUTTONS'
}

export const setColumns = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_COLUMNS
})

export const setRows = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_ROWS
})

export const setHighlights = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_HIGHLIGHT
})

export const setTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_TABLE
})

export const setShowTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_TABLE
})

export const setShowButtons = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_BUTTONS
})
