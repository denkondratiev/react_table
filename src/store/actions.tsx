import {
  ACTION_TYPES,
  Params,
  RowsParams,
  CellsParams,
  NewRowsParams
} from './types'

export const setParams = (params: Params) => ({
  payload: params,
  type: ACTION_TYPES.SET_PARAMS
})

export const setTable = (params: Array<string>) => ({
  payload: params,
  type: ACTION_TYPES.SET_TABLE
})

export const setRows = (params: RowsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_ROWS
})

export const setCells = (params: CellsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_CELLS
})

export const setShowButtons = (params: boolean) => ({
  payload: params,
  type: ACTION_TYPES.SHOW_BUTTONS
})

export const increment = (params: string) => ({
  payload: params,
  type: ACTION_TYPES.INCREMENT
})

export const setNewRow = (params: NewRowsParams) => ({
  payload: params,
  type: ACTION_TYPES.ADD_ROW
})

export const removeRow = (params: any) => ({
  payload: params,
  type: ACTION_TYPES.REMOVE_ROW
})
