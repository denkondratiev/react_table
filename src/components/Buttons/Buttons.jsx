import React from 'react'
import store from '../../store/reducers'
import { connect } from 'react-redux'
import { setRows, setTable } from '../../store/actions'
import { randomGenerate } from '../../helpers/helpers'
import { ButtonsShape } from '../../helpers/shapes'
import './Buttons.css'

const Buttons = (props) => {
  const { showButtons, rows, columns, table } = props

  const deleteRow = (event) => {
    if (rows > 1) {
      store.dispatch(setRows(rows - 1))
      store.dispatch(setTable(table.splice(0, table.length - columns)))
    }
  }

  const addRow = (event) => {
    store.dispatch(setRows(rows + 1))

    const newRow = [...Array(columns).keys()].map(randomGenerate)

    store.dispatch(setTable([...table, ...newRow]))
  }

  return (
    <>
      {showButtons && (
        <div className="buttons">
          <button
            type="button"
            className="btn btn-dark"
            onClick={deleteRow}
          >
            Delete row -
          </button>
          <button
            type="button"
            className="btn btn-dark"
            onClick={addRow}
          >
            Add row +
          </button>
        </div>
      )}
    </>
  )
}

Buttons.propTypes = ButtonsShape.isRequired

const mapStateToProps = state => ({
  showButtons: state.showButtons,
  rows: state.rows,
  columns: state.columns,
  table: state.table
})

export default connect(mapStateToProps)(Buttons)
