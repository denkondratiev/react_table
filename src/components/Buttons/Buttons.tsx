import React from 'react'
import { generateTable } from '../../helpers/generate'
import { ButtonDelete } from '../ButtonDelete/ButtonDelete'
import { ButtonAdd } from '../ButtonAdd/ButtonAdd'
import { getColumnsAmount, getTable, getShowButtons } from '../../helpers/selectors'
import './Buttons.css'
import { setNewRow, removeRow } from '../../store/actions'
import { useDispatch, useSelector } from 'react-redux'

const Buttons: React.FC = () => {

  const dispatch = useDispatch()

  const columnsAmount = useSelector(getColumnsAmount)
  const showButtons = useSelector(getShowButtons)
  const table = useSelector(getTable)

  const addRow = (): void => {
    const { table, rows, cells } = generateTable(1, columnsAmount)

    dispatch(setNewRow({ table, rows, cells}))
  }

  const deleteRow = (): void => {
    const lastRowKey = table[table.length - 1]

    dispatch(removeRow({ lastRowKey, columnsAmount }))
  }

  return (
    <>
      {showButtons && (
        <div className="buttons">
          <ButtonDelete deleteRow={deleteRow} />
          <ButtonAdd addRow={addRow} />
        </div>
      )}
    </>
  )
}

export default Buttons
