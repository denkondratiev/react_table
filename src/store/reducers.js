import { createStore } from 'redux'
import { ACTION_TYPES } from './actions'

function rootReducer (state, action) {
  const {
    SET_COLUMNS,
    SET_ROWS,
    SET_HIGHLIGHTS,
    SET_TABLE,
    SET_SHOW_TABLE,
    SET_SHOW_BUTTONS,
    INCREMENT,
    SET_LIGHT_CELLS
  } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE: {
      return {
        ...state,
        table: action.payload
      }
    }

    case SET_COLUMNS: {
      return {
        ...state,
        columns: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_ROWS: {
      return {
        ...state,
        rows: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_HIGHLIGHTS: {
      return {
        ...state,
        highlights: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_SHOW_TABLE: {
      return {
        ...state,
        showTable: action.payload
      }
    }

    case SET_SHOW_BUTTONS: {
      return {
        ...state,
        showButtons: action.payload
      }
    }

    case INCREMENT: {
      state.table.splice(action.payload.id, 1, [...state.table[action.payload.id]
        .map((item, index) => index !== action.payload.cellIndex
          ? item
          : {
            ...state.table[action.payload.id][action.payload.cellIndex],
            amount: state.table[action.payload.id][action.payload.cellIndex].amount + 1
          })
      ])
      const newTable = [...state.table]
      return {
        ...state,
        table: newTable
      }
    }

    case SET_LIGHT_CELLS: {
      // const amount = action.payload.amount
      // const highlights = action.payload.highlights
      const rowId = action.payload.id
      const cellIndex = action.payload.cellIndex
      // const col = state.table.flat().length / state.table.length
      console.log(rowId, cellIndex)
      break

      // const lightArray = [...state.table].flat().map((item, index) => ({
      //   ...item,
      //   rowIndex: index,
      //   difference: Math.abs(item.amount - amount)
      // })).sort((a, b) => a.difference - b.difference).slice(0, highlights + 1)

      // state.table.map(cell => (
      //   lightArray.some(lightCell => lightCell.cellId === cell.cellId)
      //     ? { ...cell, isLight: true }
      //     : cell
      // ))

      // return {
      //   ...state,
      //   table: newTable
      // }
    }

    default:
      return state
  }
}

const initialState = {
  rows: '',
  columns: '',
  highlights: '',
  table: [],
  showTable: false,
  showButtons: false
}

const store = createStore(
  rootReducer,
  initialState,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

export default store
