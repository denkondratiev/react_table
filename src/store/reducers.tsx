import { combineReducers } from 'redux'
import { ACTION_TYPES } from './actions'
import { Action, Params, RowsParams, CellsParams } from './actions'

const paramsReducer = (state: Params = {}, action: Action): Params => {
  const { SET_PARAMS } = ACTION_TYPES

  switch (action.type) {
    case SET_PARAMS:
      return action.payload
    default:
      return state
  }
}

function tableReducer (state: Array<string> = [], action: Action): Array<string> {
  const { SET_TABLE, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE:
      return action.payload
    case ADD_ROW:
      return [...state, action.payload.table]
    case REMOVE_ROW:
      return state.filter(id => id !== action.payload.lastRow)
    default:
      return state
  }
}

const rowsReducer = (state: RowsParams = {}, action: Action): RowsParams => {
  const { SET_ROWS, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_ROWS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.rows }
    case REMOVE_ROW:
      delete state[action.payload.lastRow]
      return state
    default:
      return state
  }
}


const cellsReducer = (state: CellsParams = {}, action: Action): CellsParams => {
  const { SET_CELLS, INCREMENT, ADD_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_CELLS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.cells }
    case INCREMENT:
      return {
        ...state,
        [action.payload]: {
          ...state[action.payload],
          amount: state[action.payload].amount + 1
        }
      }
    default:
      return state
  }
}

const buttonsReducer = (state: boolean = false, action: Action): boolean => {
  const { SHOW_BUTTONS } = ACTION_TYPES

  switch (action.type) {
    case SHOW_BUTTONS:
      return action.payload
    default:
      return state
  }
}

export const rootReducer = combineReducers({
  params: paramsReducer,
  table: tableReducer,
  rows: rowsReducer,
  cells: cellsReducer,
  buttons: buttonsReducer
})
