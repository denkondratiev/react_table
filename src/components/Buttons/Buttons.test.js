import React from 'react'
import { Provider } from 'react-redux'
import Buttons from '../Buttons/Buttons'
import { create } from 'react-test-renderer'
import store from '../../store/store'

it('should no render when false buttons', () => {
  const tree = create(
    <Provider store={store}>
      <Buttons />
    </Provider>
  )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
