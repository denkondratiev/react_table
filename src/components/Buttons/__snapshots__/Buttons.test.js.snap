// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should no render when false buttons 1`] = `null`;
